File: src/android/plugin/google/maps/PluginCircle.java
package plugin.google.maps;

import com.google.android.libraries.maps.GoogleMap;
import com.google.android.libraries.maps.model.Circle;
import com.google.android.libraries.maps.model.CircleOptions;
import com.google.android.libraries.maps.model.LatLng;
import com.google.android.libraries.maps.model.LatLngBounds;
import com.google.android.libraries.maps.model.Marker;

import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.concurrent.ConcurrentHashMap;

public class PluginCircle extends MyPlugin implements IOverlayPlugin {

  private PluginMap pluginMap;
  public final ConcurrentHashMap<String, MetaCircle> objects = new ConcurrentHashMap<String, MetaCircle>();

  /**
   * Create circle
   */
  public void create(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    final CircleOptions circleOptions = new CircleOptions();
    int color;
    final JSONObject properties = new JSONObject();

    JSONObject opts = args.getJSONObject(2);
    final String hashCode = args.getString(3);
    final String circleId = "circle_" + hashCode;
    final MetaCircle meta = new MetaCircle(circleId);

    if (opts.has("center")) {
      JSONObject center = opts.getJSONObject("center");
      circleOptions.center(new LatLng(center.getDouble("lat"), center.getDouble("lng")));
    }
    if (opts.has("radius")) {
      circleOptions.radius(opts.getDouble("radius"));
    }
    if (opts.has("strokeColor")) {
      color = PluginUtil.parsePluginColor(opts.getJSONArray("strokeColor"));
      circleOptions.strokeColor(color);
    }
    if (opts.has("fillColor")) {
      color = PluginUtil.parsePluginColor(opts.getJSONArray("fillColor"));
      circleOptions.fillColor(color);
    }
    if (opts.has("strokeWidth")) {
      circleOptions.strokeWidth((int)(opts.getDouble("strokeWidth") * density));
    }
    if (opts.has("visible")) {
      meta.isVisible = opts.getBoolean("visible");
      circleOptions.visible(meta.isVisible);
    }
    if (opts.has("zIndex")) {
      circleOptions.zIndex(opts.getInt("zIndex"));
    }
    if (opts.has("clickable")) {
      meta.isClickable = opts.getBoolean("clickable");
      properties.put("isClickable", meta.isClickable);
    } else {
      properties.put("isClickable", true);
    }
    properties.put("isVisible", circleOptions.isVisible());

    // Since this plugin provide own click detection,
    // disable default clickable feature.
    circleOptions.clickable(false);
    objects.put(circleId, meta);

    activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        Circle circle = pluginMap.getGoogleMap().addCircle(circleOptions);
        circle.setTag(circleId);

        meta.circle = circle;
        meta.properties = properties;
        meta.bounds = PluginUtil.getBoundsFromCircle(circleOptions.getCenter(), circleOptions.getRadius());

        JSONObject result = new JSONObject();
        try {
          result.put("hashCode", hashCode);
          result.put("__pgmId", circleId);
          callbackContext.success(result);
        } catch (JSONException e) {
          e.printStackTrace();
          callbackContext.error(e.getMessage() + "");
        }
      }
    });

  }

  @Override
  public void setPluginMap(PluginMap pluginMap) {
    this.pluginMap = pluginMap;
  }

  public PluginMap getMapInstance(String mapId) {
    return (PluginMap) CordovaGoogleMaps.viewPlugins.get(mapId);
  }
  public PluginCircle getInstance(String mapId) {
    PluginMap mapInstance = getMapInstance(mapId);
    return (PluginCircle) mapInstance.plugins.get(String.format("%s-circle", mapId));
  }

  /**
   * set center
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setCenter(final JSONArray args, final CallbackContext callbackContext) throws JSONException {

    String mapId = args.getString(0);
    String circleId = args.getString(1);
    JSONObject params = args.getJSONObject(2);

    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    double lat = params.getDouble("lat");
    double lng = params.getDouble("lng");
    LatLng center = new LatLng(lat, lng);

    // Recalculate the circle bounds
    meta.bounds = PluginUtil.getBoundsFromCircle(center, meta.circle.getRadius());

    meta.circle.setCenter(center);
    callbackContext.success();
  }

  /**
   * set fill color
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setFillColor(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    int color = PluginUtil.parsePluginColor(args.getJSONArray(2));
    meta.circle.setFillColor(color);
    callbackContext.success();
  }

  /**
   * set stroke color
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setStrokeColor(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    int color = PluginUtil.parsePluginColor(args.getJSONArray(2));
    meta.circle.setStrokeColor(color);
    callbackContext.success();
  }

  /**
   * set stroke width
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setStrokeWidth(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    float width = (float)(args.getDouble(2) * density);
    meta.circle.setStrokeWidth(width);
    callbackContext.success();
  }

  /**
   * set radius
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setRadius(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    double radius = args.getDouble(2);
    meta.circle.setRadius(radius);

    meta.bounds = PluginUtil.getBoundsFromCircle(meta.circle.getCenter(), meta.circle.getRadius());
    callbackContext.success();
  }

  /**
   * set z-index
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setZIndex(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    float zIndex = (float)args.getDouble(2);
    meta.circle.setZIndex(zIndex);
    callbackContext.success();
  }


  /**
   * Set visibility for the object
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setVisible(JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    boolean isVisible = args.getBoolean(2);
    meta.isVisible = isVisible;
    meta.circle.setVisible(isVisible);
    meta.properties.put("isVisible", isVisible);
    callbackContext.success();
  }

  /**
   * Set clickable for the object
   */
  @PgmPluginMethod
  public void setClickable(JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.get(circleId);

    boolean clickable = args.getBoolean(2);
    meta.isClickable = clickable;
    meta.properties.put("isClickable", clickable);
    callbackContext.success();
  }


  /**
   * Remove the circle
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void remove(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String circleId = args.getString(1);
    PluginCircle instance = getInstance(mapId);
    MetaCircle meta = instance.objects.remove(circleId);
    meta.circle.remove();
    callbackContext.success();
  }
}

File: src/browser/PluginCircle.js



var utils = require('cordova/utils'),
  event = require('cordova-plugin-googlemaps.event'),
  BaseClass = require('cordova-plugin-googlemaps.BaseClass');

function PluginCircle(pluginMap) {
  var self = this;
  BaseClass.apply(self);
  Object.defineProperty(self, 'pluginMap', {
    value: pluginMap,
    writable: false
  });
}

utils.extend(PluginCircle, BaseClass);

PluginCircle.prototype._create = function(onSuccess, onError, args) {
  var self = this,
    map = self.pluginMap.get('map'),
    circleId = 'circle_' + args[2],
    pluginOptions = args[1];

  var circleOpts = {
    'overlayId': circleId,
    'map': map
  };

  circleOpts.center = pluginOptions.center;
  circleOpts.radius = pluginOptions.radius;

  if (Array.isArray(pluginOptions.strokeColor)) {
    circleOpts.strokeColor = 'rgb(' + pluginOptions.strokeColor[0] + ',' + pluginOptions.strokeColor[1] + ',' + pluginOptions.strokeColor[2] + ')';
    circleOpts.strokeOpacity = pluginOptions.strokeColor[3] / 256;
  }
  if (Array.isArray(pluginOptions.fillColor)) {
    circleOpts.fillColor = 'rgb(' + pluginOptions.fillColor[0] + ',' + pluginOptions.fillColor[1] + ',' + pluginOptions.fillColor[2] + ')';
    circleOpts.fillOpacity = pluginOptions.fillColor[3] / 256;
  }
  if ('width' in pluginOptions) {
    circleOpts.strokeWeight = pluginOptions.width;
  }
  if ('zIndex' in pluginOptions) {
    circleOpts.zIndex = pluginOptions.zIndex;
  }
  if ('visible' in pluginOptions) {
    circleOpts.visible = pluginOptions.visible;
  }
  if ('clickable' in pluginOptions) {
    circleOpts.clickable = pluginOptions.clickable;
  }

  var circle = new google.maps.Circle(circleOpts);
  circle.addListener('click', function(polyMouseEvt) {
    self._onCircleEvent.call(self, circle, polyMouseEvt);
  });

  self.pluginMap.objects[circleId] = circle;

  onSuccess({
    'id': circleId
  });
};

PluginCircle.prototype.setCenter = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var lat = args[1];
  var lng = args[2];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    circle.setCenter(new google.maps.LatLng(lat, lng));
  }
  onSuccess();
};

PluginCircle.prototype.setFillColor = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var fillColor = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {

    if (Array.isArray(fillColor)) {
      circle.setOptions({
        'fillColor': 'rgb(' + fillColor[0] + ',' + fillColor[1] + ',' + fillColor[2] + ')',
        'fillOpacity': fillColor[3] / 256
      });
    }
  }
  onSuccess();
};

PluginCircle.prototype.setStrokeColor = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var strokeColor = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {

    if (Array.isArray(strokeColor)) {
      circle.setOptions({
        'strokeColor': 'rgb(' + strokeColor[0] + ',' + strokeColor[1] + ',' + strokeColor[2] + ')',
        'strokeOpacity': strokeColor[3] / 256
      });
    }
  }
  onSuccess();
};

PluginCircle.prototype.setStrokeWidth = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var width = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    circle.setOptions({
      'strokeWeight': width
    });
  }
  onSuccess();
};

PluginCircle.prototype.setRadius = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var radius = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    circle.setRadius(radius);
  }
  onSuccess();
};

PluginCircle.prototype.setZIndex = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var zIndex = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    circle.setOptions({
      'zIndex': zIndex
    });
  }
  onSuccess();
};

PluginCircle.prototype.setVisible = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var visible = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    circle.setVisible(visible === true);
  }
  onSuccess();
};

PluginCircle.prototype.setClickable = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var clickable = args[1];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    circle.setOptions({
      'clickable': clickable === true
    });
  }
  onSuccess();
};

PluginCircle.prototype.remove = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var circle = self.pluginMap.objects[overlayId];
  if (circle) {
    google.maps.event.clearInstanceListeners(circle);
    circle.setMap(null);
    circle = undefined;
    self.pluginMap.objects[overlayId] = undefined;
    delete self.pluginMap.objects[overlayId];
  }
  onSuccess();
};

PluginCircle.prototype._onCircleEvent = function(circle, polyMouseEvt) {
  var self = this,
    mapId = self.pluginMap.id;
  if (mapId in plugin.google.maps) {
    plugin.google.maps[mapId]({
      'evtName': event.CIRCLE_CLICK,
      'callback': '_onOverlayEvent',
      'args': [circle.overlayId, new plugin.google.maps.LatLng(polyMouseEvt.latLng.lat(), polyMouseEvt.latLng.lng())]
    });
  }

};
module.exports = PluginCircle;
